articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,130 +38,466 @@ Result result = session.run(queryString);
+`:param username => "LilRob' OR 1=1 WITH true AS hacked MATCH (all) DETACH DELETE all; //"`
 
-For example, there are some things in Cypher that cannot be parameterized, such as node labels and relation types.
-There are some APOC Procs that can help (and should be used if so), but aside from these, the only option may be to append into a query string.
+[source,cypher]
+----
+CREATE (u:User) SET u.username = $username;
+----
+
+Now that an account is made, Little Robby Tables logs in and goes to the settings to change his username.
+The database retrieves his current username and uses client-side string concatenation to build a query to update it.
+
+`query = "MATCH (u:User) WHERE u.username = '" + username + "' SET u.username = $newUsername;"`
+
+This query is executed as:
+
+[source,cypher]
+----
+MATCH (u:User) WHERE u.username = 'LilRob' OR 1=1 WITH true AS hacked MATCH (all) DETACH DELETE all; //' SET u.username = $newUsername;
+----
 
-In these cases, it is extremely important to sanitize your inputs, removing quote or delimiter characters (depending on their context of use) that would allow input to break out of the context within which you're trying to use it.
+The malicious code is now run, and all users are deleted!
+This is why sanitization should continue to be used, even when the input doesn't appear to be coming directly from a user.
 
-In these cases it is better to sanitize input in your own code at the client level, as there are many utilities across various languages for input sanitization, and it makes sense to address it at that level rather than lower down at the database itself.
+== Role Based Privileges
+
+=== Principle of Least Privilege
+
+The principle of least privilege is the idea that a program or user should have the bare minimum of privileges needed to perform their function.
+This means that, for example, if your application is only reading data, then it should have read only access to that data.
+The benefit of this is that in the case of a Cypher Injection attack, an injected query is not able to manipulate the data,
+as the role executing the hijacked query is limited in its functionality.
+With Neo4j, a range of fine-grained access control is available, which can add another layer of protection in case of an injection attack.
+See https://neo4j.com/docs/operations-manual/current/authentication-authorization/access-control/[here] for more information on role based privileges in Neo4j.
+
+== Importing Data
+
+Not all inputs can be submitted as parameters.
+Maybe some malicious input made it into a CSV file for processing.
+A CSV of the names of new students for the year, for example.
+
+[source,cypher]
+----
+LOAD CSV WITH HEADERS FROM "file:///students_2021.csv" AS row
+CREATE (s:Student)
+SET s.year = 2021, s.name = row.student_name
+----
+
+Is this vulnerable to Little Robby Labels?
+
+No, it is not. Cypher injection is still impossible here, even if parameters aren't being used.
```

Comment:
I don't know where you would use this parameter if not for the filepath. If it reference the filepath, you could definitely still do a cypher injection (if it's written on a single line), e.g:
LOAD CSV WITH HEADERS FROM "file:///students_2021.csv" as row <do whatever>//AS row CREATE (s:Student) SET s.year = 2021, s.name = row.student_name
 Otherwise please add this in the explanation :)

Reply:
I think this example makes sense still, it isn't talking about the filepath, it is talking about row.student_name. It is making it clear that it doesn't matter what is *inside* the file as it can't affect the query.

Reply:
The part that I'm confused about is that you can't parameterize row.student_name at all in currenct Cypher.

Reply:
Would it help if I wrote instead:

No, it is not. Cypher injection is still impossible here, even though parameters can't be used for the row data.

Reply:
Yes, I think that would help :)